dimod/binary/pybqm.py

```diff
@@ -0,0 +1,324 @@
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import copy
+
+from collections.abc import Collection, KeysView, Callable
+from functools import reduce
+from typing import Any, Dict, Iterator, Optional, Tuple, Mapping
+
+import numpy as np
+
+try:
+    from numpy.typing import ArrayLike, DTypeLike
+except ImportError:
+    ArrayLike = Any
+    DTypeLike = Any
+
+from dimod.sampleset import as_samples
+from dimod.typing import Variable, VartypeLike
+from dimod.utilities import iter_safe_relabels
+from dimod.vartypes import as_vartype, Vartype
+
+
+class pyBQM:
+    """A pure-python BQM implementation for handling arbitrary bias types."""
+    def __init__(self, vartype: VartypeLike):
+        self._adj: Dict[Variable, Dict[Variable, Any]] = dict()
+        self._vartype = as_vartype(vartype)
+
+        self.offset = 0.0
+
+    @property
+    def dtype(self) -> np.dtype:
+        return np.dtype('O')
+
+    @property
+    def variables(self) -> Collection:
+        return KeysView(self._adj)
+
+    @property
+    def vartype(self):
+        return self._vartype
+
+    def add_linear(self, v: Variable, bias: Any):
+        self._adj.setdefault(v, dict())
+        self._adj[v][v] = self._adj[v].get(v, 0) + bias
+
+    def add_linear_equality_constraint(self, *args, **kwargs):
+        raise NotImplementedError  # defer to caller
+
+    def add_linear_from_array(self, linear: ArrayLike):
+        for v, bias in enumerate(np.asarray(linear)):
+            self.add_linear(v, bias)
+
+    def add_quadratic(self, u: Variable, v: Variable, bias: Any):
+        if u == v:
+            raise ValueError(f"{u!r} cannot have an interaction with itself")
+
+        self.add_variable(u)
+        self.add_variable(v)
+
+        zero = type(bias)(0)
+
+        self._adj[u][v] = self._adj[v][u] = bias + self._adj[v].get(u, zero)
+
+    def add_quadratic_from_dense(self, quadratic: ArrayLike):
+        quadratic = np.asarray(quadratic)
+        if quadratic.shape[0] != quadratic.shape[1]:
+            raise ValueError("quadratic must be a square matrix")
+
+        num_variables = quadratic.shape[0]
+
+        for u in range(num_variables):
+            for v in range(num_variables):
+                if u == v:
+                    self.add_linear(u, quadratic[u, v])
+                elif quadratic[u, v]:
+                    self.add_quadratic(u, v, quadratic[u, v])
+
+    def add_variable(self, v: Optional[Variable] = None,
+                     bias: Any = 0) -> Variable:
+        if v is None:
+            # match the behaviour of dimod.Variables
+            adj = self._adj
+
+            v = len(adj)
+
+            if v in adj:
```

Comment:
Given you are checking all the elements of adj anyways, you could just do v = max(adj) + 1;

Reply:
yes, though we want to "fill in" missing values so try to keep it as close to index-labelled as possible.